src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -59,8 +73,15 @@ internal static partial class DependentTypeFinder
 
         private static readonly RelatedTypeCache s_typeToImmediatelyDerivedClassesMap = new RelatedTypeCache();
         private static readonly RelatedTypeCache s_typeToTransitivelyDerivedClassesMap = new RelatedTypeCache();
-        private static readonly RelatedTypeCache s_typeToTransitivelyImplementingStructuresClassesAndInterfacesMap = new RelatedTypeCache();
-        private static readonly RelatedTypeCache s_typeToImmediatelyDerivedAndImplementingTypesMap = new RelatedTypeCache();
```

Comment:
the data we store now is much more explicit and single-dimensional than before.  Specifically, we now store:

1. classes to their derived classes (immediate and transitive).
2. interfaces to their derived interfaces (immediate and transitive).
3. interface to their implementing types (immediate and transitive).

With this info, all other operations can be built up.